File: projects/pfinal-chat-mat/src/main/java/com/example/chat_mat/services/ChannelService.java
package com.example.chat_mat.services;

import com.example.chat_mat.dtos.ChannelDTO;
import com.example.chat_mat.entities.*;
import com.example.chat_mat.repositories.*;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

import java.util.List;

@Service
@RequiredArgsConstructor
public class ChannelService {
    private final ChannelRepository channelRepository;
    private final UserRepository userRepository;
    private final ChannelMembershipRepository channelMembershipRepository;
    private final RoleRepository roleRepository;

    public ChannelDTO createChannel(Integer userId, ChannelDTO channelRequest) {
        var user = userRepository.findById(userId).orElseThrow(() ->
                new IllegalArgumentException("User not found"));

        var channel = new Channel();
        channel.setName(channelRequest.getName());
        channel.setIsDeleted(false);
        channel = channelRepository.save(channel);

        var ownerRole = roleRepository.findByRoleName("OWNER").orElseThrow(() ->
                new IllegalArgumentException("Role 'OWNER' not found"));

        var membership = new ChannelMembership();
        membership.setChannel(channel);
        membership.setUser(user);
        membership.setRole(ownerRole);
        channelMembershipRepository.save(membership);

        return mapToChannelDTO(channel);
    }

    public void editChannelName(Integer ownerId, Integer channelId, String newName) {
        var channel = channelRepository.findById(channelId).orElseThrow(() ->
                new IllegalArgumentException("Channel not found"));

        var ownerRole = roleRepository.findByRoleName("OWNER").orElseThrow(() ->
                new IllegalArgumentException("Role 'OWNER' not found"));

        var isOwner = channelMembershipRepository.existsByChannelIdAndUserIdAndRoleId(
                channelId, ownerId, ownerRole.getId());
        if (!isOwner) {
            throw new IllegalArgumentException("User is not owner of the channel");
        }

        if (newName == null || newName.trim().isEmpty()) {
            throw new IllegalArgumentException("Channel name cannot be empty");
        }

        channel.setName(newName.trim());
        channelRepository.save(channel);
    }

    public List<ChannelDTO> getUserChannels(Integer userId) {
         var user = userRepository.findById(userId).orElseThrow(() ->
                new IllegalArgumentException("User not found"));

        return channelMembershipRepository.findByUserId(userId)
                .stream()
                .filter(membership -> !membership.getChannel().getIsDeleted())
                .map(membership -> mapToChannelDTO(membership.getChannel(), membership.getRole().getId()))
                .toList();
    }

    public void deleteChannel(Integer userId, Integer channelId) {
        var channel = channelRepository.findById(channelId).orElseThrow(() ->
                new IllegalArgumentException("Channel not found"));

        var ownerRole = roleRepository.findByRoleName("OWNER").orElseThrow(() ->
                new IllegalArgumentException("Role 'OWNER' not found"));
        var ownerRoleId = ownerRole.getId();

        var isOwner = channelMembershipRepository.existsByChannelIdAndUserIdAndRoleId(
                channelId, userId, ownerRoleId);
        if (!isOwner) {
            throw new IllegalArgumentException("User is not the owner of the channel");
        }

        channel.setIsDeleted(true);
        channelRepository.save(channel);
    }

    public void addGuestToChannel(Integer userId, Integer channelId, Integer guestId) {
        var channel = channelRepository.findById(channelId).orElseThrow(() ->
                new IllegalArgumentException("Channel not found"));

        var ownerRole = roleRepository.findByRoleName("OWNER").orElseThrow(() ->
                new IllegalArgumentException("Role 'OWNER' not found"));
        var adminRole = roleRepository.findByRoleName("ADMIN").orElseThrow(() ->
                new IllegalArgumentException("Role 'ADMIN' not found"));

        var hasPermission =
                channelMembershipRepository.existsByChannelIdAndUserIdAndRoleId(channelId, userId, ownerRole.getId())
                || channelMembershipRepository.existsByChannelIdAndUserIdAndRoleId(channelId, userId, adminRole.getId());

        if (!hasPermission) {
            throw new IllegalArgumentException("Only the channel owner or admin can add guests");
        }

        var guest = userRepository.findById(guestId).orElseThrow(() ->
                new IllegalArgumentException("Guest user not found"));

        var isAlreadyMember = channelMembershipRepository.existsByChannelIdAndUserId(channelId, guestId);
        if (isAlreadyMember) {
            throw new IllegalArgumentException("User is already a member of the channel");
        }

        var guestRole = roleRepository.findByRoleName("GUEST").orElseThrow(() ->
                new IllegalArgumentException("Role 'GUEST' not found"));

        var newMembership = new ChannelMembership();
        newMembership.setChannel(channel);
        newMembership.setUser(guest);
        newMembership.setRole(guestRole);
        channelMembershipRepository.save(newMembership);
    }

    public void removeGuestFromChannel(Integer ownerId, Integer channelId, Integer guestId) {
        var channel = channelRepository.findById(channelId).orElseThrow(() ->
                new IllegalArgumentException("Channel not found"));

        var ownerRole = roleRepository.findByRoleName("OWNER").orElseThrow(() ->
                new IllegalArgumentException("Role 'OWNER' not found"));

        var isOwner = channelMembershipRepository.existsByChannelIdAndUserIdAndRoleId(
                channelId, ownerId, ownerRole.getId());
        if (!isOwner) {
            throw new IllegalArgumentException("Only the channel owner can remove users");
        }

        var guestRole = roleRepository.findByRoleName("GUEST").orElseThrow(() ->
                new IllegalArgumentException("Role 'GUEST' not found"));

        var guestMembership = channelMembershipRepository
                .findByChannelIdAndUserIdAndRoleId(channelId, guestId, guestRole.getId())
                .orElseThrow(() -> new IllegalArgumentException("User is not a guest in this channel"));

        channelMembershipRepository.delete(guestMembership);
    }

    public void promoteToAdmin(Integer ownerId, Integer channelId, Integer guestId) {
        var channel = channelRepository.findById(channelId).orElseThrow(() ->
                new IllegalArgumentException("Channel not found"));

        var ownerRole = roleRepository.findByRoleName("OWNER").orElseThrow(() ->
                new IllegalArgumentException("Role 'OWNER' not found"));

        boolean isOwner = channelMembershipRepository.existsByChannelIdAndUserIdAndRoleId(
                channelId, ownerId, ownerRole.getId());
        if (!isOwner) {
            throw new IllegalArgumentException("Only the channel owner can promote users");
        }

        var guestRole = roleRepository.findByRoleName("GUEST").orElseThrow(() ->
                new IllegalArgumentException("Role 'GUEST' not found"));

        var membership = channelMembershipRepository.findByChannelIdAndUserIdAndRoleId(channelId, guestId, guestRole.getId())
                .orElseThrow(() -> new IllegalArgumentException("User is not a GUEST in this channel"));

        var isPromotingOwner = channelMembershipRepository.existsByChannelIdAndUserIdAndRoleId(channelId, guestId, ownerRole.getId());
        if (isPromotingOwner) {
            throw new IllegalArgumentException("Cannot change the role of OWNER");
        }

        var adminRole = roleRepository.findByRoleName("ADMIN").orElseThrow(() ->
                new IllegalArgumentException("Role 'ADMIN' not found"));
        membership.setRole(adminRole);
        channelMembershipRepository.save(membership);
    }

    private ChannelDTO mapToChannelDTO(Channel channel) {
        ChannelDTO dto = new ChannelDTO();
        dto.setId(channel.getId());
        dto.setName(channel.getName());
        dto.setIsDeleted(channel.getIsDeleted());
        return dto;
    }

    private ChannelDTO mapToChannelDTO(Channel channel, Integer roleId) {
        ChannelDTO dto = new ChannelDTO();
        dto.setId(channel.getId());
        dto.setName(channel.getName());
        dto.setIsDeleted(channel.getIsDeleted());
        dto.setRoleId(roleId);
        return dto;
    }
}

File: projects/pfinal-chat-mat-fe/src/services/channelService.js
import axios from "axios";
import { globalConstants } from "../utils/constants";

const baseUrl = globalConstants.BASE_URL;

export async function getChannels(userId) {
  try {
    const response = await axios.get(`${baseUrl}/users/${userId}/channels`);
    if (response.data.code === 200) {
      return response.data.data;
    } else {
      throw new Error("Failed to fetch channels");
    }
  } catch (error) {
    console.log(error);
    return [];
  }
}

export async function getChannelMembers(channelId) {
  try {
    const response = await axios.get(
      `${baseUrl}/channels/${channelId}/members`
    );
    if (response.data.code === 200) {
      return response.data.data.map((member) => {
        return {
          ...member.user,
          role: member.role,
        };
      });
    }
    throw new Error("Failed to fetch channel members");
  } catch (error) {
    console.log(error);
    return [];
  }
}

export async function removeGuest(channelId, ownerId, guestId) {
  try {
    const response = await axios.delete(
      `${baseUrl}/channels/${channelId}/remove-guest`,
      null,
      {
        params: { ownerId, guestId },
      }
    );
    return response.data;
  } catch (error) {
    console.log(error);
    return null;
  }
}

export async function promoteToAdmin(channelId, ownerId, guestId) {
  try {
    await axios.post(
      `${baseUrl}/channels/${channelId}/promote-to-admin`,
      null,
      {
        params: { ownerId, guestId },
      }
    );
    throw new Error("Failed to promote guest to admin");
  } catch (error) {
    console.log(error);
  }
}

export async function addGuest(channelId, userId, guestId) {
  try {
    await axios.post(`${baseUrl}/channels/${channelId}/add-guest`, null, {
      params: { userId, guestId },
    });

    throw new Error("Failed to add guest");
  } catch (error) {
    console.log(error);
  }
}

export async function changeChannelName(channelId, ownerId, newName) {
  try {
    const response = await axios.put(
      `${baseUrl}/channels`,
      { id: channelId, name: newName },
      { params: { ownerId } }
    );

    if (response.data.code === 200) {
      return response.data.message;
    }
    throw new Error("Failed to change channel name");
  } catch (error) {
    console.log(error);
    return null;
  }
}

export async function createChannel(userId, channelName) {
  try {
    const response = await axios.post(`${baseUrl}/channels?userId=${userId}`, {
      name: channelName,
    });
    if (response.data.code === 201) {
      return response.data.data;
    }
  } catch (error) {
    console.error("Error creating channel:", error);
    throw error;
  }
}
